### PR TITLE
Create SAI-Proposal-Generic-Counters.md

### PR DESCRIPTION
**Pull request by marian-pritsak**

### CODE REVIEW

File: inc/saifdb.h [hunk 1]
```diff
@@ -165,6 +165,18 @@ typedef enum _sai_fdb_entry_attr_t
...
+     * @brief Attach a list of counters
+     *
+     * When it is empty, then packet hits won't be counted
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_COUNTER
+     * @default empty
+     */
+    SAI_FDB_ENTRY_ATTR_COUNTER_ID_LIST,
```
Comment (lguohan) [inline inc/saifdb.h]: do we want list here? I thought we discussed this last time, if we have list then we need to get know how many counters we can attach to an entry.

Comment (marian-pritsak) [reply inc/saifdb.h]: Changed to a single ID

File: inc/saicounter.h [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    /** Get/set hit packets count [uint64_t] */
+    SAI_COUNTER_STAT_PACKETS = 0x00000000,
+
+    /** Get/set hit bytes count [uint64_t] */
+    SAI_COUNTER_STAT_BYTES = 0x00000001,
+
+    /** Custom range base value */
+    SAI_COUNTER_STAT_CUSTOM_RANGE_BASE = 0x10000000
+
+} sai_counter_stat_t;
```
[the remark this answers is not in the document]
Comment (marian-pritsak) [reply inc/saicounter.h]: The enum name is according to the convention, like sai_port_stat_t, sai_queue_stat_t, etc.
It is used in the functions below sai_get_counter_stats_fn and sai_get_counter_stats_ext_fn for counter_ids parameter. You may notice that the counter_ids type is sai_stat_id_t, and not sai_counter_stat_t, but this is also according to the template for get_stat functions.

File: inc/saicounter.h [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+ * @param[in] counter_ids Specifies the array of counter ids
+ * @param[out] counters Array of resulting counter values.
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_get_counter_stats_fn)(
+        _In_ sai_object_id_t counter_id,
+        _In_ uint32_t number_of_counters,
+        _In_ const sai_stat_id_t *counter_ids,
+        _Out_ uint64_t *counters);
```
[the remark this answers is not in the document]
Comment (marian-pritsak) [reply inc/saicounter.h]: I've added a usage example to the proposal. Let me know if it helps.

File: inc/saicounter.h [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+ * @param[out] counters Array of resulting counter values.
+ *
+ * @return #SAI_STATUS_SUCCESS on success, failure status code on error
+ */
+typedef sai_status_t (*sai_get_counter_stats_ext_fn)(
+        _In_ sai_object_id_t counter_id,
+        _In_ uint32_t number_of_counters,
+        _In_ const sai_stat_id_t *counter_ids,
+        _In_ sai_stats_mode_t mode,
+        _Out_ uint64_t *counters);
```
[the remark this answers is not in the document]
Comment (marian-pritsak) [reply inc/saicounter.h]: I've added a usage example to the proposal. Let me know if it helps.

File: doc/SAI-Proposal-Generic-Counters.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        _In_ const sai_stat_id_t *stat_ids,
+        _In_ sai_stats_mode_t stats_mode,
+        _Out_ uint64_t *stats);
+```
+
+## Counter object behavior
+* Counter can be attached/detached multiple times during it's lifetime.
+* Detaching a counter from an object does not clear it's stats values.
+* Attaching a counter to an object does not clear it's stats values.
+* Counter can be attached to multiple objects at the same time.
```
[the remark this answers is not in the document]
Comment (marian-pritsak) [reply doc/SAI-Proposal-Generic-Counters.md]: Counting is cumulative. One counter can be used to aggregate values from a set of objects. If the desired behaviour is to count per object, then a new counter should be created for each object.

File: doc/SAI-Proposal-Generic-Counters.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
+# Generic SAI Counters and Resources
+
+Title       | Critical Resource Monitor
```
Comment (tushar-ty) [inline doc/SAI-Proposal-Generic-Counters.md]: Title should be fixed.

File: doc/SAI-Proposal-Generic-Counters.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+     * @brief Attach a list of counters
+     *
+     * When it is empty, then packet hits won't be counted
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_COUNTER
+     * @default empty
+     */
+    SAI_ROUTE_ENTRY_ATTR_COUNTER_ID_LIST,
```
Comment (tushar-ty) [inline doc/SAI-Proposal-Generic-Counters.md]: I see a list here and not in the example - whats the final choice ?

File: doc/SAI-Proposal-Generic-Counters.md (hunk 6 above)
Comment (marian-pritsak) [reply doc/SAI-Proposal-Generic-Counters.md]: Done

Comment (marian-pritsak) [reply doc/SAI-Proposal-Generic-Counters.md]: Done

File: doc/SAI-Proposal-Generic-Counters.md (hunk 7 above)
Comment (marian-pritsak) [reply doc/SAI-Proposal-Generic-Counters.md]: The final choice is to start with a single id - it's simpler and object query can be implemented with today's API

File: doc/SAI-Proposal-Generic-Counters.md [hunk 8]
```diff
@@ -39,16 +39,16 @@ typedef enum _sai_route_entry_attr_t
...
-     * @brief Attach a list of counters
+     * @brief Attach a counter
      *
      * When it is empty, then packet hits won't be counted
      *
-     * @type sai_object_list_t
+     * @type sai_object_id_t
      * @flags CREATE_AND_SET
      * @objects SAI_OBJECT_TYPE_COUNTER
      * @default empty
```
Comment (kcudnik) [inline doc/SAI-Proposal-Generic-Counters.md]: null instead of empty ?

Comment (marian-pritsak) [reply doc/SAI-Proposal-Generic-Counters.md]: Yes, updated

**Pull request merged**